// @switchboard-xyz/aptos.js/lib/generated/types/AggregatorResultsConfig.js
import * as types from "./index.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
import { BN } from "@switchboard-xyz/common"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class AggregatorResultsConfig {
    varianceThreshold;
    forceReportPeriod;
    minJobResults;
    expiration;
    constructor(fields) {
        this.varianceThreshold = fields.varianceThreshold;
        this.forceReportPeriod = fields.forceReportPeriod;
        this.minJobResults = fields.minJobResults;
        this.expiration = fields.expiration;
    }
    toJSON() {
        return {
            varianceThreshold: this.varianceThreshold.toJSON(),
            forceReportPeriod: this.forceReportPeriod.toString(),
            minJobResults: this.minJobResults.toString(),
            expiration: this.expiration.toString(),
        };
    }
    static fromJSON(obj) {
        return new AggregatorResultsConfig({
            varianceThreshold: types.SwitchboardDecimal.fromJSON(obj.varianceThreshold),
            forceReportPeriod: new BN(obj.forceReportPeriod),
            minJobResults: new BN(obj.minJobResults),
            expiration: new BN(obj.expiration),
        });
    }
    toMoveStruct() {
        return {
            variance_threshold: this.varianceThreshold.toMoveStruct(),
            force_report_period: this.forceReportPeriod.toString(),
            min_job_results: this.minJobResults.toString(),
            expiration: this.expiration.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new AggregatorResultsConfig({
            varianceThreshold: types.SwitchboardDecimal.fromMoveStruct(obj.variance_threshold),
            forceReportPeriod: new BN(obj.force_report_period),
            minJobResults: new BN(obj.min_job_results),
            expiration: new BN(obj.expiration),
        });
    }
}
