// @switchboard-xyz/aptos.js/lib/generated/types/AggregatorCrankEvictionEvent.js
import { BN } from "@switchboard-xyz/common"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class AggregatorCrankEvictionEvent {
    crankAddress;
    aggregatorAddress;
    reason;
    timestamp;
    constructor(fields) {
        this.crankAddress = fields.crankAddress;
        this.aggregatorAddress = fields.aggregatorAddress;
        this.reason = fields.reason;
        this.timestamp = fields.timestamp;
    }
    toJSON() {
        return {
            crankAddress: this.crankAddress.toString(),
            aggregatorAddress: this.aggregatorAddress.toString(),
            reason: this.reason.toString(),
            timestamp: this.timestamp.toString(),
        };
    }
    static fromJSON(obj) {
        return new AggregatorCrankEvictionEvent({
            crankAddress: obj.crankAddress,
            aggregatorAddress: obj.aggregatorAddress,
            reason: new BN(obj.reason),
            timestamp: new BN(obj.timestamp),
        });
    }
    toMoveStruct() {
        return {
            crank_address: this.crankAddress.toString(),
            aggregator_address: this.aggregatorAddress.toString(),
            reason: this.reason.toString(),
            timestamp: this.timestamp.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new AggregatorCrankEvictionEvent({
            crankAddress: obj.crank_address,
            aggregatorAddress: obj.aggregator_address,
            reason: new BN(obj.reason),
            timestamp: new BN(obj.timestamp),
        });
    }
}
