// @switchboard-xyz/aptos.js/lib/generated/types/AggregatorUpdateEvent.js
import * as types from "./index.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class AggregatorUpdateEvent {
    aggregatorAddress;
    oldValue;
    newValue;
    constructor(fields) {
        this.aggregatorAddress = fields.aggregatorAddress;
        this.oldValue = fields.oldValue;
        this.newValue = fields.newValue;
    }
    toJSON() {
        return {
            aggregatorAddress: this.aggregatorAddress.toString(),
            oldValue: this.oldValue.toJSON(),
            newValue: this.newValue.toJSON(),
        };
    }
    static fromJSON(obj) {
        return new AggregatorUpdateEvent({
            aggregatorAddress: obj.aggregatorAddress,
            oldValue: types.SwitchboardDecimal.fromJSON(obj.oldValue),
            newValue: types.SwitchboardDecimal.fromJSON(obj.newValue),
        });
    }
    toMoveStruct() {
        return {
            aggregator_address: this.aggregatorAddress.toString(),
            old_value: this.oldValue.toMoveStruct(),
            new_value: this.newValue.toMoveStruct(),
        };
    }
    static fromMoveStruct(obj) {
        return new AggregatorUpdateEvent({
            aggregatorAddress: obj.aggregator_address,
            oldValue: types.SwitchboardDecimal.fromMoveStruct(obj.old_value),
            newValue: types.SwitchboardDecimal.fromMoveStruct(obj.new_value),
        });
    }
}
