// @switchboard-xyz/aptos.js/lib/generated/types/AggregatorSaveResultEvent.js
import * as types from "./index.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class AggregatorSaveResultEvent {
    aggregatorAddress;
    oracleKey;
    value;
    constructor(fields) {
        this.aggregatorAddress = fields.aggregatorAddress;
        this.oracleKey = fields.oracleKey;
        this.value = fields.value;
    }
    toJSON() {
        return {
            aggregatorAddress: this.aggregatorAddress.toString(),
            oracleKey: this.oracleKey.toString(),
            value: this.value.toJSON(),
        };
    }
    static fromJSON(obj) {
        return new AggregatorSaveResultEvent({
            aggregatorAddress: obj.aggregatorAddress,
            oracleKey: obj.oracleKey,
            value: types.SwitchboardDecimal.fromJSON(obj.value),
        });
    }
    toMoveStruct() {
        return {
            aggregator_address: this.aggregatorAddress.toString(),
            oracle_key: this.oracleKey.toString(),
            value: this.value.toMoveStruct(),
        };
    }
    static fromMoveStruct(obj) {
        return new AggregatorSaveResultEvent({
            aggregatorAddress: obj.aggregator_address,
            oracleKey: obj.oracle_key,
            value: types.SwitchboardDecimal.fromMoveStruct(obj.value),
        });
    }
}
