// @switchboard-xyz/aptos.js/lib/generated/types/AggregatorUpdateEvent.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AggregatorUpdateEvent = void 0;
const types = __importStar(require("./index.cjs")); // eslint-disable-line @typescript-eslint/no-unused-vars
class AggregatorUpdateEvent {
    aggregatorAddress;
    oldValue;
    newValue;
    constructor(fields) {
        this.aggregatorAddress = fields.aggregatorAddress;
        this.oldValue = fields.oldValue;
        this.newValue = fields.newValue;
    }
    toJSON() {
        return {
            aggregatorAddress: this.aggregatorAddress.toString(),
            oldValue: this.oldValue.toJSON(),
            newValue: this.newValue.toJSON(),
        };
    }
    static fromJSON(obj) {
        return new AggregatorUpdateEvent({
            aggregatorAddress: obj.aggregatorAddress,
            oldValue: types.SwitchboardDecimal.fromJSON(obj.oldValue),
            newValue: types.SwitchboardDecimal.fromJSON(obj.newValue),
        });
    }
    toMoveStruct() {
        return {
            aggregator_address: this.aggregatorAddress.toString(),
            old_value: this.oldValue.toMoveStruct(),
            new_value: this.newValue.toMoveStruct(),
        };
    }
    static fromMoveStruct(obj) {
        return new AggregatorUpdateEvent({
            aggregatorAddress: obj.aggregator_address,
            oldValue: types.SwitchboardDecimal.fromMoveStruct(obj.old_value),
            newValue: types.SwitchboardDecimal.fromMoveStruct(obj.new_value),
        });
    }
}
exports.AggregatorUpdateEvent = AggregatorUpdateEvent;
