// @switchboard-xyz/aptos.js/lib/generated/types/AggregatorConfig.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AggregatorConfig = void 0;
const types = __importStar(require("./index.cjs")); // eslint-disable-line @typescript-eslint/no-unused-vars
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class AggregatorConfig {
    queueAddr;
    batchSize;
    minOracleResults;
    minUpdateDelaySeconds;
    historyLimit;
    varianceThreshold;
    forceReportPeriod;
    minJobResults;
    expiration;
    crankAddr;
    crankDisabled;
    crankRowCount;
    nextAllowedUpdateTime;
    consecutiveFailureCount;
    startAfter;
    constructor(fields) {
        this.queueAddr = fields.queueAddr;
        this.batchSize = fields.batchSize;
        this.minOracleResults = fields.minOracleResults;
        this.minUpdateDelaySeconds = fields.minUpdateDelaySeconds;
        this.historyLimit = fields.historyLimit;
        this.varianceThreshold = fields.varianceThreshold;
        this.forceReportPeriod = fields.forceReportPeriod;
        this.minJobResults = fields.minJobResults;
        this.expiration = fields.expiration;
        this.crankAddr = fields.crankAddr;
        this.crankDisabled = fields.crankDisabled;
        this.crankRowCount = fields.crankRowCount;
        this.nextAllowedUpdateTime = fields.nextAllowedUpdateTime;
        this.consecutiveFailureCount = fields.consecutiveFailureCount;
        this.startAfter = fields.startAfter;
    }
    toJSON() {
        return {
            queueAddr: this.queueAddr.toString(),
            batchSize: this.batchSize.toString(),
            minOracleResults: this.minOracleResults.toString(),
            minUpdateDelaySeconds: this.minUpdateDelaySeconds.toString(),
            historyLimit: this.historyLimit.toString(),
            varianceThreshold: this.varianceThreshold.toJSON(),
            forceReportPeriod: this.forceReportPeriod.toString(),
            minJobResults: this.minJobResults.toString(),
            expiration: this.expiration.toString(),
            crankAddr: this.crankAddr.toString(),
            crankDisabled: this.crankDisabled,
            crankRowCount: this.crankRowCount.toString(),
            nextAllowedUpdateTime: this.nextAllowedUpdateTime.toString(),
            consecutiveFailureCount: this.consecutiveFailureCount.toString(),
            startAfter: this.startAfter.toString(),
        };
    }
    static fromJSON(obj) {
        return new AggregatorConfig({
            queueAddr: obj.queueAddr,
            batchSize: new common_1.BN(obj.batchSize),
            minOracleResults: new common_1.BN(obj.minOracleResults),
            minUpdateDelaySeconds: new common_1.BN(obj.minUpdateDelaySeconds),
            historyLimit: new common_1.BN(obj.historyLimit),
            varianceThreshold: types.SwitchboardDecimal.fromJSON(obj.varianceThreshold),
            forceReportPeriod: new common_1.BN(obj.forceReportPeriod),
            minJobResults: new common_1.BN(obj.minJobResults),
            expiration: new common_1.BN(obj.expiration),
            crankAddr: obj.crankAddr,
            crankDisabled: obj.crankDisabled,
            crankRowCount: new common_1.BN(obj.crankRowCount),
            nextAllowedUpdateTime: new common_1.BN(obj.nextAllowedUpdateTime),
            consecutiveFailureCount: new common_1.BN(obj.consecutiveFailureCount),
            startAfter: new common_1.BN(obj.startAfter),
        });
    }
    toMoveStruct() {
        return {
            queue_addr: this.queueAddr.toString(),
            batch_size: this.batchSize.toString(),
            min_oracle_results: this.minOracleResults.toString(),
            min_update_delay_seconds: this.minUpdateDelaySeconds.toString(),
            history_limit: this.historyLimit.toString(),
            variance_threshold: this.varianceThreshold.toMoveStruct(),
            force_report_period: this.forceReportPeriod.toString(),
            min_job_results: this.minJobResults.toString(),
            expiration: this.expiration.toString(),
            crank_addr: this.crankAddr.toString(),
            crank_disabled: this.crankDisabled,
            crank_row_count: this.crankRowCount.toString(),
            next_allowed_update_time: this.nextAllowedUpdateTime.toString(),
            consecutive_failure_count: this.consecutiveFailureCount.toString(),
            start_after: this.startAfter.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new AggregatorConfig({
            queueAddr: obj.queue_addr,
            batchSize: new common_1.BN(obj.batch_size),
            minOracleResults: new common_1.BN(obj.min_oracle_results),
            minUpdateDelaySeconds: new common_1.BN(obj.min_update_delay_seconds),
            historyLimit: new common_1.BN(obj.history_limit),
            varianceThreshold: types.SwitchboardDecimal.fromMoveStruct(obj.variance_threshold),
            forceReportPeriod: new common_1.BN(obj.force_report_period),
            minJobResults: new common_1.BN(obj.min_job_results),
            expiration: new common_1.BN(obj.expiration),
            crankAddr: obj.crank_addr,
            crankDisabled: obj.crank_disabled,
            crankRowCount: new common_1.BN(obj.crank_row_count),
            nextAllowedUpdateTime: new common_1.BN(obj.next_allowed_update_time),
            consecutiveFailureCount: new common_1.BN(obj.consecutive_failure_count),
            startAfter: new common_1.BN(obj.start_after),
        });
    }
}
exports.AggregatorConfig = AggregatorConfig;
