// @switchboard-xyz/aptos.js/lib/generated/types/AggregatorReadConfig.js
import { BN } from "@switchboard-xyz/common"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class AggregatorReadConfig {
    readCharge;
    rewardEscrow;
    readWhitelist;
    limitReadsToWhitelist;
    constructor(fields) {
        this.readCharge = fields.readCharge;
        this.rewardEscrow = fields.rewardEscrow;
        this.readWhitelist = fields.readWhitelist;
        this.limitReadsToWhitelist = fields.limitReadsToWhitelist;
    }
    toJSON() {
        return {
            readCharge: this.readCharge.toString(),
            rewardEscrow: this.rewardEscrow.toString(),
            readWhitelist: this.readWhitelist.map((item) => item.toString()),
            limitReadsToWhitelist: this.limitReadsToWhitelist,
        };
    }
    static fromJSON(obj) {
        return new AggregatorReadConfig({
            readCharge: new BN(obj.readCharge),
            rewardEscrow: obj.rewardEscrow,
            readWhitelist: obj.readWhitelist,
            limitReadsToWhitelist: obj.limitReadsToWhitelist,
        });
    }
    toMoveStruct() {
        return {
            read_charge: this.readCharge.toString(),
            reward_escrow: this.rewardEscrow.toString(),
            read_whitelist: this.readWhitelist.map((item) => item.toString()),
            limit_reads_to_whitelist: this.limitReadsToWhitelist,
        };
    }
    static fromMoveStruct(obj) {
        return new AggregatorReadConfig({
            readCharge: new BN(obj.read_charge),
            rewardEscrow: obj.reward_escrow,
            readWhitelist: obj.read_whitelist,
            limitReadsToWhitelist: obj.limit_reads_to_whitelist,
        });
    }
}
