// @switchboard-xyz/aptos.js/lib/generated/types/AggregatorResultsConfig.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AggregatorResultsConfig = void 0;
const types = __importStar(require("./index.cjs")); // eslint-disable-line @typescript-eslint/no-unused-vars
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class AggregatorResultsConfig {
    varianceThreshold;
    forceReportPeriod;
    minJobResults;
    expiration;
    constructor(fields) {
        this.varianceThreshold = fields.varianceThreshold;
        this.forceReportPeriod = fields.forceReportPeriod;
        this.minJobResults = fields.minJobResults;
        this.expiration = fields.expiration;
    }
    toJSON() {
        return {
            varianceThreshold: this.varianceThreshold.toJSON(),
            forceReportPeriod: this.forceReportPeriod.toString(),
            minJobResults: this.minJobResults.toString(),
            expiration: this.expiration.toString(),
        };
    }
    static fromJSON(obj) {
        return new AggregatorResultsConfig({
            varianceThreshold: types.SwitchboardDecimal.fromJSON(obj.varianceThreshold),
            forceReportPeriod: new common_1.BN(obj.forceReportPeriod),
            minJobResults: new common_1.BN(obj.minJobResults),
            expiration: new common_1.BN(obj.expiration),
        });
    }
    toMoveStruct() {
        return {
            variance_threshold: this.varianceThreshold.toMoveStruct(),
            force_report_period: this.forceReportPeriod.toString(),
            min_job_results: this.minJobResults.toString(),
            expiration: this.expiration.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new AggregatorResultsConfig({
            varianceThreshold: types.SwitchboardDecimal.fromMoveStruct(obj.variance_threshold),
            forceReportPeriod: new common_1.BN(obj.force_report_period),
            minJobResults: new common_1.BN(obj.min_job_results),
            expiration: new common_1.BN(obj.expiration),
        });
    }
}
exports.AggregatorResultsConfig = AggregatorResultsConfig;
