// @switchboard-xyz/aptos.js/lib/generated/types/AggregatorHistoryRow.js
import * as types from "./index.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
import { BN } from "@switchboard-xyz/common"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class AggregatorHistoryRow {
    value;
    timestamp;
    roundId;
    constructor(fields) {
        this.value = fields.value;
        this.timestamp = fields.timestamp;
        this.roundId = fields.roundId;
    }
    toJSON() {
        return {
            value: this.value.toJSON(),
            timestamp: this.timestamp.toString(),
            roundId: this.roundId.toString(),
        };
    }
    static fromJSON(obj) {
        return new AggregatorHistoryRow({
            value: types.SwitchboardDecimal.fromJSON(obj.value),
            timestamp: new BN(obj.timestamp),
            roundId: new BN(obj.roundId),
        });
    }
    toMoveStruct() {
        return {
            value: this.value.toMoveStruct(),
            timestamp: this.timestamp.toString(),
            round_id: this.roundId.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new AggregatorHistoryRow({
            value: types.SwitchboardDecimal.fromMoveStruct(obj.value),
            timestamp: new BN(obj.timestamp),
            roundId: new BN(obj.round_id),
        });
    }
}
