// @switchboard-xyz/aptos.js/lib/generated/types/AggregatorHistoryData.js
import * as types from "./index.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
import { BN } from "@switchboard-xyz/common"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class AggregatorHistoryData {
    history;
    historyWriteIdx;
    constructor(fields) {
        this.history = fields.history;
        this.historyWriteIdx = fields.historyWriteIdx;
    }
    toJSON() {
        return {
            history: this.history.map((item) => item.toJSON()),
            historyWriteIdx: this.historyWriteIdx.toString(),
        };
    }
    static fromJSON(obj) {
        return new AggregatorHistoryData({
            history: obj.history.map((item) => types.AggregatorHistoryRow.fromJSON(item)),
            historyWriteIdx: new BN(obj.historyWriteIdx),
        });
    }
    toMoveStruct() {
        return {
            history: this.history.map((item) => item.toMoveStruct()),
            history_write_idx: this.historyWriteIdx.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new AggregatorHistoryData({
            history: obj.history.map((item) => types.AggregatorHistoryRow.fromMoveStruct(item)),
            historyWriteIdx: new BN(obj.history_write_idx),
        });
    }
}
