// @switchboard-xyz/aptos.js/lib/generated/types/AggregatorAddJobParams.js
export class AggregatorAddJobParams {
    aggregatorAddr;
    jobAddr;
    weight;
    constructor(fields) {
        this.aggregatorAddr = fields.aggregatorAddr;
        this.jobAddr = fields.jobAddr;
        this.weight = fields.weight;
    }
    toJSON() {
        return {
            aggregatorAddr: this.aggregatorAddr.toString(),
            jobAddr: this.jobAddr.toString(),
            weight: this.weight,
        };
    }
    static fromJSON(obj) {
        return new AggregatorAddJobParams({
            aggregatorAddr: obj.aggregatorAddr,
            jobAddr: obj.jobAddr,
            weight: obj.weight,
        });
    }
    toMoveStruct() {
        return {
            aggregator_addr: this.aggregatorAddr.toString(),
            job_addr: this.jobAddr.toString(),
            weight: this.weight,
        };
    }
    static fromMoveStruct(obj) {
        return new AggregatorAddJobParams({
            aggregatorAddr: obj.aggregator_addr,
            jobAddr: obj.job_addr,
            weight: obj.weight,
        });
    }
}
