// @switchboard-xyz/aptos.js/lib/generated/types/AggregatorOpenRoundEvent.js
export class AggregatorOpenRoundEvent {
    aggregatorAddress;
    oracleKeys;
    jobKeys;
    constructor(fields) {
        this.aggregatorAddress = fields.aggregatorAddress;
        this.oracleKeys = fields.oracleKeys;
        this.jobKeys = fields.jobKeys;
    }
    toJSON() {
        return {
            aggregatorAddress: this.aggregatorAddress.toString(),
            oracleKeys: this.oracleKeys.map((item) => item.toString()),
            jobKeys: this.jobKeys.map((item) => item.toString()),
        };
    }
    static fromJSON(obj) {
        return new AggregatorOpenRoundEvent({
            aggregatorAddress: obj.aggregatorAddress,
            oracleKeys: obj.oracleKeys,
            jobKeys: obj.jobKeys,
        });
    }
    toMoveStruct() {
        return {
            aggregator_address: this.aggregatorAddress.toString(),
            oracle_keys: this.oracleKeys.map((item) => item.toString()),
            job_keys: this.jobKeys.map((item) => item.toString()),
        };
    }
    static fromMoveStruct(obj) {
        return new AggregatorOpenRoundEvent({
            aggregatorAddress: obj.aggregator_address,
            oracleKeys: obj.oracle_keys,
            jobKeys: obj.job_keys,
        });
    }
}
