// @switchboard-xyz/aptos.js/lib/generated/types/AggregatorOpenRoundParams.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AggregatorOpenRoundParams = void 0;
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class AggregatorOpenRoundParams {
    aggregatorAddr;
    jitter;
    constructor(fields) {
        this.aggregatorAddr = fields.aggregatorAddr;
        this.jitter = fields.jitter;
    }
    toJSON() {
        return {
            aggregatorAddr: this.aggregatorAddr.toString(),
            jitter: this.jitter.toString(),
        };
    }
    static fromJSON(obj) {
        return new AggregatorOpenRoundParams({
            aggregatorAddr: obj.aggregatorAddr,
            jitter: new common_1.BN(obj.jitter),
        });
    }
    toMoveStruct() {
        return {
            aggregator_addr: this.aggregatorAddr.toString(),
            jitter: this.jitter.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new AggregatorOpenRoundParams({
            aggregatorAddr: obj.aggregator_addr,
            jitter: new common_1.BN(obj.jitter),
        });
    }
}
exports.AggregatorOpenRoundParams = AggregatorOpenRoundParams;
