// @switchboard-xyz/aptos.js/lib/generated/types/Crank.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Crank = void 0;
const types = __importStar(require("./index.cjs")); // eslint-disable-line @typescript-eslint/no-unused-vars
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class Crank {
    heap;
    queueAddr;
    createdAt;
    jitterModifier;
    features;
    _ebuf;
    constructor(fields) {
        this.heap = fields.heap;
        this.queueAddr = fields.queueAddr;
        this.createdAt = fields.createdAt;
        this.jitterModifier = fields.jitterModifier;
        this.features = fields.features;
        this._ebuf = fields._ebuf;
    }
    toJSON() {
        return {
            heap: this.heap.map((item) => item.toJSON()),
            queueAddr: this.queueAddr.toString(),
            createdAt: this.createdAt.toString(),
            jitterModifier: this.jitterModifier.toString(),
            features: this.features.map((item) => item),
            _ebuf: [...this._ebuf],
        };
    }
    static fromJSON(obj) {
        return new Crank({
            heap: obj.heap.map((item) => types.CrankRow.fromJSON(item)),
            queueAddr: obj.queueAddr,
            createdAt: new common_1.BN(obj.createdAt),
            jitterModifier: new common_1.BN(obj.jitterModifier),
            features: obj.features.map((item) => item),
            _ebuf: new Uint8Array(obj._ebuf),
        });
    }
    toMoveStruct() {
        return {
            heap: this.heap.map((item) => item.toMoveStruct()),
            queue_addr: this.queueAddr.toString(),
            created_at: this.createdAt.toString(),
            jitter_modifier: this.jitterModifier.toString(),
            features: this.features.map((item) => item),
            _ebuf: Buffer.from(this._ebuf).toString("hex"),
        };
    }
    static fromMoveStruct(obj) {
        return new Crank({
            heap: obj.heap.map((item) => types.CrankRow.fromMoveStruct(item)),
            queueAddr: obj.queue_addr,
            createdAt: new common_1.BN(obj.created_at),
            jitterModifier: new common_1.BN(obj.jitter_modifier),
            features: obj.features.map((item) => item),
            _ebuf: typeof obj._ebuf === "string"
                ? new Uint8Array(Buffer.from(obj._ebuf.slice(2), "hex"))
                : new Uint8Array(obj._ebuf),
        });
    }
}
exports.Crank = Crank;
