// @switchboard-xyz/aptos.js/lib/generated/types/CrankPopParams.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CrankPopParams = void 0;
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class CrankPopParams {
    crankAddr;
    popIdx;
    constructor(fields) {
        this.crankAddr = fields.crankAddr;
        this.popIdx = fields.popIdx;
    }
    toJSON() {
        return {
            crankAddr: this.crankAddr.toString(),
            popIdx: this.popIdx.toString(),
        };
    }
    static fromJSON(obj) {
        return new CrankPopParams({
            crankAddr: obj.crankAddr,
            popIdx: new common_1.BN(obj.popIdx),
        });
    }
    toMoveStruct() {
        return {
            crank_addr: this.crankAddr.toString(),
            pop_idx: this.popIdx.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new CrankPopParams({
            crankAddr: obj.crank_addr,
            popIdx: new common_1.BN(obj.pop_idx),
        });
    }
}
exports.CrankPopParams = CrankPopParams;
