// @switchboard-xyz/aptos.js/lib/generated/types/CrankRow.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CrankRow = void 0;
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class CrankRow {
    aggregatorAddr;
    timestamp;
    constructor(fields) {
        this.aggregatorAddr = fields.aggregatorAddr;
        this.timestamp = fields.timestamp;
    }
    toJSON() {
        return {
            aggregatorAddr: this.aggregatorAddr.toString(),
            timestamp: this.timestamp.toString(),
        };
    }
    static fromJSON(obj) {
        return new CrankRow({
            aggregatorAddr: obj.aggregatorAddr,
            timestamp: new common_1.BN(obj.timestamp),
        });
    }
    toMoveStruct() {
        return {
            aggregator_addr: this.aggregatorAddr.toString(),
            timestamp: this.timestamp.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new CrankRow({
            aggregatorAddr: obj.aggregator_addr,
            timestamp: new common_1.BN(obj.timestamp),
        });
    }
}
exports.CrankRow = CrankRow;
