// @switchboard-xyz/aptos.js/lib/generated/types/FeedPermissionRevokedEvent.js
import { BN } from "@switchboard-xyz/common"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class FeedPermissionRevokedEvent {
    aggregatorAddress;
    timestamp;
    constructor(fields) {
        this.aggregatorAddress = fields.aggregatorAddress;
        this.timestamp = fields.timestamp;
    }
    toJSON() {
        return {
            aggregatorAddress: this.aggregatorAddress.toString(),
            timestamp: this.timestamp.toString(),
        };
    }
    static fromJSON(obj) {
        return new FeedPermissionRevokedEvent({
            aggregatorAddress: obj.aggregatorAddress,
            timestamp: new BN(obj.timestamp),
        });
    }
    toMoveStruct() {
        return {
            aggregator_address: this.aggregatorAddress.toString(),
            timestamp: this.timestamp.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new FeedPermissionRevokedEvent({
            aggregatorAddress: obj.aggregator_address,
            timestamp: new BN(obj.timestamp),
        });
    }
}
