// @switchboard-xyz/aptos.js/lib/generated/types/JobConfigParams.js
export class JobConfigParams {
    name;
    metadata;
    authority;
    data;
    constructor(fields) {
        this.name = fields.name;
        this.metadata = fields.metadata;
        this.authority = fields.authority;
        this.data = fields.data;
    }
    toJSON() {
        return {
            name: [...this.name],
            metadata: [...this.metadata],
            authority: this.authority.toString(),
            data: [...this.data],
        };
    }
    static fromJSON(obj) {
        return new JobConfigParams({
            name: new Uint8Array(obj.name),
            metadata: new Uint8Array(obj.metadata),
            authority: obj.authority,
            data: new Uint8Array(obj.data),
        });
    }
    toMoveStruct() {
        return {
            name: Buffer.from(this.name).toString("hex"),
            metadata: Buffer.from(this.metadata).toString("hex"),
            authority: this.authority.toString(),
            data: Buffer.from(this.data).toString("hex"),
        };
    }
    static fromMoveStruct(obj) {
        return new JobConfigParams({
            name: typeof obj.name === "string"
                ? new Uint8Array(Buffer.from(obj.name.slice(2), "hex"))
                : new Uint8Array(obj.name),
            metadata: typeof obj.metadata === "string"
                ? new Uint8Array(Buffer.from(obj.metadata.slice(2), "hex"))
                : new Uint8Array(obj.metadata),
            authority: obj.authority,
            data: typeof obj.data === "string"
                ? new Uint8Array(Buffer.from(obj.data.slice(2), "hex"))
                : new Uint8Array(obj.data),
        });
    }
}
