// @switchboard-xyz/aptos.js/lib/generated/types/LeaseWithdrawParams.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LeaseWithdrawParams = void 0;
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class LeaseWithdrawParams {
    aggregatorAddr;
    queueAddr;
    amount;
    constructor(fields) {
        this.aggregatorAddr = fields.aggregatorAddr;
        this.queueAddr = fields.queueAddr;
        this.amount = fields.amount;
    }
    toJSON() {
        return {
            aggregatorAddr: this.aggregatorAddr.toString(),
            queueAddr: this.queueAddr.toString(),
            amount: this.amount.toString(),
        };
    }
    static fromJSON(obj) {
        return new LeaseWithdrawParams({
            aggregatorAddr: obj.aggregatorAddr,
            queueAddr: obj.queueAddr,
            amount: new common_1.BN(obj.amount),
        });
    }
    toMoveStruct() {
        return {
            aggregator_addr: this.aggregatorAddr.toString(),
            queue_addr: this.queueAddr.toString(),
            amount: this.amount.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new LeaseWithdrawParams({
            aggregatorAddr: obj.aggregator_addr,
            queueAddr: obj.queue_addr,
            amount: new common_1.BN(obj.amount),
        });
    }
}
exports.LeaseWithdrawParams = LeaseWithdrawParams;
