// @switchboard-xyz/aptos.js/lib/generated/types/LeaseExtendParams.js
import { BN } from "@switchboard-xyz/common"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class LeaseExtendParams {
    aggregatorAddr;
    loadAmount;
    constructor(fields) {
        this.aggregatorAddr = fields.aggregatorAddr;
        this.loadAmount = fields.loadAmount;
    }
    toJSON() {
        return {
            aggregatorAddr: this.aggregatorAddr.toString(),
            loadAmount: this.loadAmount.toString(),
        };
    }
    static fromJSON(obj) {
        return new LeaseExtendParams({
            aggregatorAddr: obj.aggregatorAddr,
            loadAmount: new BN(obj.loadAmount),
        });
    }
    toMoveStruct() {
        return {
            aggregator_addr: this.aggregatorAddr.toString(),
            load_amount: this.loadAmount.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new LeaseExtendParams({
            aggregatorAddr: obj.aggregator_addr,
            loadAmount: new BN(obj.load_amount),
        });
    }
}
