// @switchboard-xyz/aptos.js/lib/generated/types/LeaseSetAuthorityparams.js
export class LeaseSetAuthorityparams {
    aggregatorAddr;
    queueAddr;
    authority;
    constructor(fields) {
        this.aggregatorAddr = fields.aggregatorAddr;
        this.queueAddr = fields.queueAddr;
        this.authority = fields.authority;
    }
    toJSON() {
        return {
            aggregatorAddr: this.aggregatorAddr.toString(),
            queueAddr: this.queueAddr.toString(),
            authority: this.authority.toString(),
        };
    }
    static fromJSON(obj) {
        return new LeaseSetAuthorityparams({
            aggregatorAddr: obj.aggregatorAddr,
            queueAddr: obj.queueAddr,
            authority: obj.authority,
        });
    }
    toMoveStruct() {
        return {
            aggregator_addr: this.aggregatorAddr.toString(),
            queue_addr: this.queueAddr.toString(),
            authority: this.authority.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new LeaseSetAuthorityparams({
            aggregatorAddr: obj.aggregator_addr,
            queueAddr: obj.queue_addr,
            authority: obj.authority,
        });
    }
}
