// @switchboard-xyz/aptos.js/lib/generated/types/LeaseWithdrawEvent.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LeaseWithdrawEvent = void 0;
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class LeaseWithdrawEvent {
    leaseAddress;
    destinationWallet;
    previousAmount;
    newAmount;
    timestamp;
    constructor(fields) {
        this.leaseAddress = fields.leaseAddress;
        this.destinationWallet = fields.destinationWallet;
        this.previousAmount = fields.previousAmount;
        this.newAmount = fields.newAmount;
        this.timestamp = fields.timestamp;
    }
    toJSON() {
        return {
            leaseAddress: this.leaseAddress.toString(),
            destinationWallet: this.destinationWallet.toString(),
            previousAmount: this.previousAmount.toString(),
            newAmount: this.newAmount.toString(),
            timestamp: this.timestamp.toString(),
        };
    }
    static fromJSON(obj) {
        return new LeaseWithdrawEvent({
            leaseAddress: obj.leaseAddress,
            destinationWallet: obj.destinationWallet,
            previousAmount: new common_1.BN(obj.previousAmount),
            newAmount: new common_1.BN(obj.newAmount),
            timestamp: new common_1.BN(obj.timestamp),
        });
    }
    toMoveStruct() {
        return {
            lease_address: this.leaseAddress.toString(),
            destination_wallet: this.destinationWallet.toString(),
            previous_amount: this.previousAmount.toString(),
            new_amount: this.newAmount.toString(),
            timestamp: this.timestamp.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new LeaseWithdrawEvent({
            leaseAddress: obj.lease_address,
            destinationWallet: obj.destination_wallet,
            previousAmount: new common_1.BN(obj.previous_amount),
            newAmount: new common_1.BN(obj.new_amount),
            timestamp: new common_1.BN(obj.timestamp),
        });
    }
}
exports.LeaseWithdrawEvent = LeaseWithdrawEvent;
