// @switchboard-xyz/aptos.js/lib/generated/types/LeaseExtendParams.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LeaseExtendParams = void 0;
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class LeaseExtendParams {
    aggregatorAddr;
    loadAmount;
    constructor(fields) {
        this.aggregatorAddr = fields.aggregatorAddr;
        this.loadAmount = fields.loadAmount;
    }
    toJSON() {
        return {
            aggregatorAddr: this.aggregatorAddr.toString(),
            loadAmount: this.loadAmount.toString(),
        };
    }
    static fromJSON(obj) {
        return new LeaseExtendParams({
            aggregatorAddr: obj.aggregatorAddr,
            loadAmount: new common_1.BN(obj.loadAmount),
        });
    }
    toMoveStruct() {
        return {
            aggregator_addr: this.aggregatorAddr.toString(),
            load_amount: this.loadAmount.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new LeaseExtendParams({
            aggregatorAddr: obj.aggregator_addr,
            loadAmount: new common_1.BN(obj.load_amount),
        });
    }
}
exports.LeaseExtendParams = LeaseExtendParams;
