// @switchboard-xyz/aptos.js/lib/generated/types/LeaseFundEvent.js
import { BN } from "@switchboard-xyz/common"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class LeaseFundEvent {
    leaseAddress;
    funder;
    amount;
    timestamp;
    constructor(fields) {
        this.leaseAddress = fields.leaseAddress;
        this.funder = fields.funder;
        this.amount = fields.amount;
        this.timestamp = fields.timestamp;
    }
    toJSON() {
        return {
            leaseAddress: this.leaseAddress.toString(),
            funder: this.funder.toString(),
            amount: this.amount.toString(),
            timestamp: this.timestamp.toString(),
        };
    }
    static fromJSON(obj) {
        return new LeaseFundEvent({
            leaseAddress: obj.leaseAddress,
            funder: obj.funder,
            amount: new BN(obj.amount),
            timestamp: new BN(obj.timestamp),
        });
    }
    toMoveStruct() {
        return {
            lease_address: this.leaseAddress.toString(),
            funder: this.funder.toString(),
            amount: this.amount.toString(),
            timestamp: this.timestamp.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new LeaseFundEvent({
            leaseAddress: obj.lease_address,
            funder: obj.funder,
            amount: new BN(obj.amount),
            timestamp: new BN(obj.timestamp),
        });
    }
}
