// @switchboard-xyz/aptos.js/lib/generated/types/OracleSlashEvent.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OracleSlashEvent = void 0;
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class OracleSlashEvent {
    aggregatorAddress;
    oracleAddress;
    amount;
    timestamp;
    constructor(fields) {
        this.aggregatorAddress = fields.aggregatorAddress;
        this.oracleAddress = fields.oracleAddress;
        this.amount = fields.amount;
        this.timestamp = fields.timestamp;
    }
    toJSON() {
        return {
            aggregatorAddress: this.aggregatorAddress.toString(),
            oracleAddress: this.oracleAddress.toString(),
            amount: this.amount.toString(),
            timestamp: this.timestamp.toString(),
        };
    }
    static fromJSON(obj) {
        return new OracleSlashEvent({
            aggregatorAddress: obj.aggregatorAddress,
            oracleAddress: obj.oracleAddress,
            amount: new common_1.BN(obj.amount),
            timestamp: new common_1.BN(obj.timestamp),
        });
    }
    toMoveStruct() {
        return {
            aggregator_address: this.aggregatorAddress.toString(),
            oracle_address: this.oracleAddress.toString(),
            amount: this.amount.toString(),
            timestamp: this.timestamp.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new OracleSlashEvent({
            aggregatorAddress: obj.aggregator_address,
            oracleAddress: obj.oracle_address,
            amount: new common_1.BN(obj.amount),
            timestamp: new common_1.BN(obj.timestamp),
        });
    }
}
exports.OracleSlashEvent = OracleSlashEvent;
