// @switchboard-xyz/aptos.js/lib/generated/types/OracleBootedEvent.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OracleBootedEvent = void 0;
class OracleBootedEvent {
    queueAddress;
    oracleAddress;
    constructor(fields) {
        this.queueAddress = fields.queueAddress;
        this.oracleAddress = fields.oracleAddress;
    }
    toJSON() {
        return {
            queueAddress: this.queueAddress.toString(),
            oracleAddress: this.oracleAddress.toString(),
        };
    }
    static fromJSON(obj) {
        return new OracleBootedEvent({
            queueAddress: obj.queueAddress,
            oracleAddress: obj.oracleAddress,
        });
    }
    toMoveStruct() {
        return {
            queue_address: this.queueAddress.toString(),
            oracle_address: this.oracleAddress.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new OracleBootedEvent({
            queueAddress: obj.queue_address,
            oracleAddress: obj.oracle_address,
        });
    }
}
exports.OracleBootedEvent = OracleBootedEvent;
