// @switchboard-xyz/aptos.js/lib/generated/types/Oracle.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Oracle = void 0;
const types = __importStar(require("./index.cjs")); // eslint-disable-line @typescript-eslint/no-unused-vars
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class Oracle {
    signerCap;
    name;
    metadata;
    authority;
    lastHeartbeat;
    queueAddr;
    numRows;
    createdAt;
    metrics;
    features;
    _ebuf;
    constructor(fields) {
        this.signerCap = fields.signerCap;
        this.name = fields.name;
        this.metadata = fields.metadata;
        this.authority = fields.authority;
        this.lastHeartbeat = fields.lastHeartbeat;
        this.queueAddr = fields.queueAddr;
        this.numRows = fields.numRows;
        this.createdAt = fields.createdAt;
        this.metrics = fields.metrics;
        this.features = fields.features;
        this._ebuf = fields._ebuf;
    }
    toJSON() {
        return {
            signerCap: this.signerCap.toJSON(),
            name: [...this.name],
            metadata: [...this.metadata],
            authority: this.authority.toString(),
            lastHeartbeat: this.lastHeartbeat.toString(),
            queueAddr: this.queueAddr.toString(),
            numRows: this.numRows,
            createdAt: this.createdAt.toString(),
            metrics: this.metrics.toJSON(),
            features: this.features.map((item) => item),
            _ebuf: [...this._ebuf],
        };
    }
    static fromJSON(obj) {
        return new Oracle({
            signerCap: types.SignerCapability.fromJSON(obj.signerCap),
            name: new Uint8Array(obj.name),
            metadata: new Uint8Array(obj.metadata),
            authority: obj.authority,
            lastHeartbeat: new common_1.BN(obj.lastHeartbeat),
            queueAddr: obj.queueAddr,
            numRows: obj.numRows,
            createdAt: new common_1.BN(obj.createdAt),
            metrics: types.OracleMetrics.fromJSON(obj.metrics),
            features: obj.features.map((item) => item),
            _ebuf: new Uint8Array(obj._ebuf),
        });
    }
    toMoveStruct() {
        return {
            signer_cap: this.signerCap.toMoveStruct(),
            name: Buffer.from(this.name).toString("hex"),
            metadata: Buffer.from(this.metadata).toString("hex"),
            authority: this.authority.toString(),
            last_heartbeat: this.lastHeartbeat.toString(),
            queue_addr: this.queueAddr.toString(),
            num_rows: this.numRows,
            created_at: this.createdAt.toString(),
            metrics: this.metrics.toMoveStruct(),
            features: this.features.map((item) => item),
            _ebuf: Buffer.from(this._ebuf).toString("hex"),
        };
    }
    static fromMoveStruct(obj) {
        return new Oracle({
            signerCap: types.SignerCapability.fromMoveStruct(obj.signer_cap),
            name: typeof obj.name === "string"
                ? new Uint8Array(Buffer.from(obj.name.slice(2), "hex"))
                : new Uint8Array(obj.name),
            metadata: typeof obj.metadata === "string"
                ? new Uint8Array(Buffer.from(obj.metadata.slice(2), "hex"))
                : new Uint8Array(obj.metadata),
            authority: obj.authority,
            lastHeartbeat: new common_1.BN(obj.last_heartbeat),
            queueAddr: obj.queue_addr,
            numRows: obj.num_rows,
            createdAt: new common_1.BN(obj.created_at),
            metrics: types.OracleMetrics.fromMoveStruct(obj.metrics),
            features: obj.features.map((item) => item),
            _ebuf: typeof obj._ebuf === "string"
                ? new Uint8Array(Buffer.from(obj._ebuf.slice(2), "hex"))
                : new Uint8Array(obj._ebuf),
        });
    }
}
exports.Oracle = Oracle;
