// @switchboard-xyz/aptos.js/lib/generated/types/OracleRewardEvent.js
import { BN } from "@switchboard-xyz/common"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class OracleRewardEvent {
    aggregatorAddress;
    oracleAddress;
    amount;
    timestamp;
    constructor(fields) {
        this.aggregatorAddress = fields.aggregatorAddress;
        this.oracleAddress = fields.oracleAddress;
        this.amount = fields.amount;
        this.timestamp = fields.timestamp;
    }
    toJSON() {
        return {
            aggregatorAddress: this.aggregatorAddress.toString(),
            oracleAddress: this.oracleAddress.toString(),
            amount: this.amount.toString(),
            timestamp: this.timestamp.toString(),
        };
    }
    static fromJSON(obj) {
        return new OracleRewardEvent({
            aggregatorAddress: obj.aggregatorAddress,
            oracleAddress: obj.oracleAddress,
            amount: new BN(obj.amount),
            timestamp: new BN(obj.timestamp),
        });
    }
    toMoveStruct() {
        return {
            aggregator_address: this.aggregatorAddress.toString(),
            oracle_address: this.oracleAddress.toString(),
            amount: this.amount.toString(),
            timestamp: this.timestamp.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new OracleRewardEvent({
            aggregatorAddress: obj.aggregator_address,
            oracleAddress: obj.oracle_address,
            amount: new BN(obj.amount),
            timestamp: new BN(obj.timestamp),
        });
    }
}
