// @switchboard-xyz/aptos.js/lib/generated/types/OracleQueueSetConfigsParams.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OracleQueueSetConfigsParams = void 0;
const types = __importStar(require("./index.cjs")); // eslint-disable-line @typescript-eslint/no-unused-vars
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class OracleQueueSetConfigsParams {
    addr;
    name;
    metadata;
    authority;
    oracleTimeout;
    reward;
    minStake;
    slashingEnabled;
    varianceToleranceMultiplier;
    feedProbationPeriod;
    consecutiveFeedFailureLimit;
    consecutiveOracleFailureLimit;
    unpermissionedFeedsEnabled;
    lockLeaseFunding;
    maxSize;
    saveConfirmationReward;
    saveReward;
    openRoundReward;
    slashingPenalty;
    constructor(fields) {
        this.addr = fields.addr;
        this.name = fields.name;
        this.metadata = fields.metadata;
        this.authority = fields.authority;
        this.oracleTimeout = fields.oracleTimeout;
        this.reward = fields.reward;
        this.minStake = fields.minStake;
        this.slashingEnabled = fields.slashingEnabled;
        this.varianceToleranceMultiplier = fields.varianceToleranceMultiplier;
        this.feedProbationPeriod = fields.feedProbationPeriod;
        this.consecutiveFeedFailureLimit = fields.consecutiveFeedFailureLimit;
        this.consecutiveOracleFailureLimit = fields.consecutiveOracleFailureLimit;
        this.unpermissionedFeedsEnabled = fields.unpermissionedFeedsEnabled;
        this.lockLeaseFunding = fields.lockLeaseFunding;
        this.maxSize = fields.maxSize;
        this.saveConfirmationReward = fields.saveConfirmationReward;
        this.saveReward = fields.saveReward;
        this.openRoundReward = fields.openRoundReward;
        this.slashingPenalty = fields.slashingPenalty;
    }
    toJSON() {
        return {
            addr: this.addr.toString(),
            name: [...this.name],
            metadata: [...this.metadata],
            authority: this.authority.toString(),
            oracleTimeout: this.oracleTimeout.toString(),
            reward: this.reward.toString(),
            minStake: this.minStake.toString(),
            slashingEnabled: this.slashingEnabled,
            varianceToleranceMultiplier: this.varianceToleranceMultiplier.toJSON(),
            feedProbationPeriod: this.feedProbationPeriod.toString(),
            consecutiveFeedFailureLimit: this.consecutiveFeedFailureLimit.toString(),
            consecutiveOracleFailureLimit: this.consecutiveOracleFailureLimit.toString(),
            unpermissionedFeedsEnabled: this.unpermissionedFeedsEnabled,
            lockLeaseFunding: this.lockLeaseFunding,
            maxSize: this.maxSize.toString(),
            saveConfirmationReward: this.saveConfirmationReward.toString(),
            saveReward: this.saveReward.toString(),
            openRoundReward: this.openRoundReward.toString(),
            slashingPenalty: this.slashingPenalty.toString(),
        };
    }
    static fromJSON(obj) {
        return new OracleQueueSetConfigsParams({
            addr: obj.addr,
            name: new Uint8Array(obj.name),
            metadata: new Uint8Array(obj.metadata),
            authority: obj.authority,
            oracleTimeout: new common_1.BN(obj.oracleTimeout),
            reward: new common_1.BN(obj.reward),
            minStake: new common_1.BN(obj.minStake),
            slashingEnabled: obj.slashingEnabled,
            varianceToleranceMultiplier: types.SwitchboardDecimal.fromJSON(obj.varianceToleranceMultiplier),
            feedProbationPeriod: new common_1.BN(obj.feedProbationPeriod),
            consecutiveFeedFailureLimit: new common_1.BN(obj.consecutiveFeedFailureLimit),
            consecutiveOracleFailureLimit: new common_1.BN(obj.consecutiveOracleFailureLimit),
            unpermissionedFeedsEnabled: obj.unpermissionedFeedsEnabled,
            lockLeaseFunding: obj.lockLeaseFunding,
            maxSize: new common_1.BN(obj.maxSize),
            saveConfirmationReward: new common_1.BN(obj.saveConfirmationReward),
            saveReward: new common_1.BN(obj.saveReward),
            openRoundReward: new common_1.BN(obj.openRoundReward),
            slashingPenalty: new common_1.BN(obj.slashingPenalty),
        });
    }
    toMoveStruct() {
        return {
            addr: this.addr.toString(),
            name: Buffer.from(this.name).toString("hex"),
            metadata: Buffer.from(this.metadata).toString("hex"),
            authority: this.authority.toString(),
            oracle_timeout: this.oracleTimeout.toString(),
            reward: this.reward.toString(),
            min_stake: this.minStake.toString(),
            slashing_enabled: this.slashingEnabled,
            variance_tolerance_multiplier: this.varianceToleranceMultiplier.toMoveStruct(),
            feed_probation_period: this.feedProbationPeriod.toString(),
            consecutive_feed_failure_limit: this.consecutiveFeedFailureLimit.toString(),
            consecutive_oracle_failure_limit: this.consecutiveOracleFailureLimit.toString(),
            unpermissioned_feeds_enabled: this.unpermissionedFeedsEnabled,
            lock_lease_funding: this.lockLeaseFunding,
            max_size: this.maxSize.toString(),
            save_confirmation_reward: this.saveConfirmationReward.toString(),
            save_reward: this.saveReward.toString(),
            open_round_reward: this.openRoundReward.toString(),
            slashing_penalty: this.slashingPenalty.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new OracleQueueSetConfigsParams({
            addr: obj.addr,
            name: typeof obj.name === "string"
                ? new Uint8Array(Buffer.from(obj.name.slice(2), "hex"))
                : new Uint8Array(obj.name),
            metadata: typeof obj.metadata === "string"
                ? new Uint8Array(Buffer.from(obj.metadata.slice(2), "hex"))
                : new Uint8Array(obj.metadata),
            authority: obj.authority,
            oracleTimeout: new common_1.BN(obj.oracle_timeout),
            reward: new common_1.BN(obj.reward),
            minStake: new common_1.BN(obj.min_stake),
            slashingEnabled: obj.slashing_enabled,
            varianceToleranceMultiplier: types.SwitchboardDecimal.fromMoveStruct(obj.variance_tolerance_multiplier),
            feedProbationPeriod: new common_1.BN(obj.feed_probation_period),
            consecutiveFeedFailureLimit: new common_1.BN(obj.consecutive_feed_failure_limit),
            consecutiveOracleFailureLimit: new common_1.BN(obj.consecutive_oracle_failure_limit),
            unpermissionedFeedsEnabled: obj.unpermissioned_feeds_enabled,
            lockLeaseFunding: obj.lock_lease_funding,
            maxSize: new common_1.BN(obj.max_size),
            saveConfirmationReward: new common_1.BN(obj.save_confirmation_reward),
            saveReward: new common_1.BN(obj.save_reward),
            openRoundReward: new common_1.BN(obj.open_round_reward),
            slashingPenalty: new common_1.BN(obj.slashing_penalty),
        });
    }
}
exports.OracleQueueSetConfigsParams = OracleQueueSetConfigsParams;
