// @switchboard-xyz/aptos.js/lib/generated/types/OracleWalletWithdrawParams.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OracleWalletWithdrawParams = void 0;
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class OracleWalletWithdrawParams {
    oracleAddr;
    queueAddr;
    amount;
    constructor(fields) {
        this.oracleAddr = fields.oracleAddr;
        this.queueAddr = fields.queueAddr;
        this.amount = fields.amount;
    }
    toJSON() {
        return {
            oracleAddr: this.oracleAddr.toString(),
            queueAddr: this.queueAddr.toString(),
            amount: this.amount.toString(),
        };
    }
    static fromJSON(obj) {
        return new OracleWalletWithdrawParams({
            oracleAddr: obj.oracleAddr,
            queueAddr: obj.queueAddr,
            amount: new common_1.BN(obj.amount),
        });
    }
    toMoveStruct() {
        return {
            oracle_addr: this.oracleAddr.toString(),
            queue_addr: this.queueAddr.toString(),
            amount: this.amount.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new OracleWalletWithdrawParams({
            oracleAddr: obj.oracle_addr,
            queueAddr: obj.queue_addr,
            amount: new common_1.BN(obj.amount),
        });
    }
}
exports.OracleWalletWithdrawParams = OracleWalletWithdrawParams;
