// @switchboard-xyz/aptos.js/lib/generated/types/OracleWithdrawEvent.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OracleWithdrawEvent = void 0;
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class OracleWithdrawEvent {
    oracleAddress;
    destinationWallet;
    previousAmount;
    newAmount;
    timestamp;
    constructor(fields) {
        this.oracleAddress = fields.oracleAddress;
        this.destinationWallet = fields.destinationWallet;
        this.previousAmount = fields.previousAmount;
        this.newAmount = fields.newAmount;
        this.timestamp = fields.timestamp;
    }
    toJSON() {
        return {
            oracleAddress: this.oracleAddress.toString(),
            destinationWallet: this.destinationWallet.toString(),
            previousAmount: this.previousAmount.toString(),
            newAmount: this.newAmount.toString(),
            timestamp: this.timestamp.toString(),
        };
    }
    static fromJSON(obj) {
        return new OracleWithdrawEvent({
            oracleAddress: obj.oracleAddress,
            destinationWallet: obj.destinationWallet,
            previousAmount: new common_1.BN(obj.previousAmount),
            newAmount: new common_1.BN(obj.newAmount),
            timestamp: new common_1.BN(obj.timestamp),
        });
    }
    toMoveStruct() {
        return {
            oracle_address: this.oracleAddress.toString(),
            destination_wallet: this.destinationWallet.toString(),
            previous_amount: this.previousAmount.toString(),
            new_amount: this.newAmount.toString(),
            timestamp: this.timestamp.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new OracleWithdrawEvent({
            oracleAddress: obj.oracle_address,
            destinationWallet: obj.destination_wallet,
            previousAmount: new common_1.BN(obj.previous_amount),
            newAmount: new common_1.BN(obj.new_amount),
            timestamp: new common_1.BN(obj.timestamp),
        });
    }
}
exports.OracleWithdrawEvent = OracleWithdrawEvent;
