// @switchboard-xyz/aptos.js/lib/generated/types/OracleQueueData.js
import { BN } from "@switchboard-xyz/common"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class OracleQueueData {
    data;
    currIdx;
    gcIdx;
    constructor(fields) {
        this.data = fields.data;
        this.currIdx = fields.currIdx;
        this.gcIdx = fields.gcIdx;
    }
    toJSON() {
        return {
            data: this.data.map((item) => item.toString()),
            currIdx: this.currIdx.toString(),
            gcIdx: this.gcIdx.toString(),
        };
    }
    static fromJSON(obj) {
        return new OracleQueueData({
            data: obj.data.map((item) => item),
            currIdx: new BN(obj.currIdx),
            gcIdx: new BN(obj.gcIdx),
        });
    }
    toMoveStruct() {
        return {
            data: this.data.map((item) => item.toString()),
            curr_idx: this.currIdx.toString(),
            gc_idx: this.gcIdx.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new OracleQueueData({
            data: obj.data.map((item) => item),
            currIdx: new BN(obj.curr_idx),
            gcIdx: new BN(obj.gc_idx),
        });
    }
}
