// @switchboard-xyz/aptos.js/lib/generated/types/OracleQueueData.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OracleQueueData = void 0;
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class OracleQueueData {
    data;
    currIdx;
    gcIdx;
    constructor(fields) {
        this.data = fields.data;
        this.currIdx = fields.currIdx;
        this.gcIdx = fields.gcIdx;
    }
    toJSON() {
        return {
            data: this.data.map((item) => item.toString()),
            currIdx: this.currIdx.toString(),
            gcIdx: this.gcIdx.toString(),
        };
    }
    static fromJSON(obj) {
        return new OracleQueueData({
            data: obj.data.map((item) => item),
            currIdx: new common_1.BN(obj.currIdx),
            gcIdx: new common_1.BN(obj.gcIdx),
        });
    }
    toMoveStruct() {
        return {
            data: this.data.map((item) => item.toString()),
            curr_idx: this.currIdx.toString(),
            gc_idx: this.gcIdx.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new OracleQueueData({
            data: obj.data.map((item) => item),
            currIdx: new common_1.BN(obj.curr_idx),
            gcIdx: new common_1.BN(obj.gc_idx),
        });
    }
}
exports.OracleQueueData = OracleQueueData;
