// @switchboard-xyz/aptos.js/lib/generated/types/OracleData.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OracleData = void 0;
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class OracleData {
    numRows;
    lastHeartbeat;
    constructor(fields) {
        this.numRows = fields.numRows;
        this.lastHeartbeat = fields.lastHeartbeat;
    }
    toJSON() {
        return {
            numRows: this.numRows,
            lastHeartbeat: this.lastHeartbeat.toString(),
        };
    }
    static fromJSON(obj) {
        return new OracleData({
            numRows: obj.numRows,
            lastHeartbeat: new common_1.BN(obj.lastHeartbeat),
        });
    }
    toMoveStruct() {
        return {
            num_rows: this.numRows,
            last_heartbeat: this.lastHeartbeat.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new OracleData({
            numRows: obj.num_rows,
            lastHeartbeat: new common_1.BN(obj.last_heartbeat),
        });
    }
}
exports.OracleData = OracleData;
