// @switchboard-xyz/aptos.js/lib/generated/types/OracleWalletInitParams.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OracleWalletInitParams = void 0;
class OracleWalletInitParams {
    oracleAddr;
    queueAddr;
    withdrawAuthority;
    constructor(fields) {
        this.oracleAddr = fields.oracleAddr;
        this.queueAddr = fields.queueAddr;
        this.withdrawAuthority = fields.withdrawAuthority;
    }
    toJSON() {
        return {
            oracleAddr: this.oracleAddr.toString(),
            queueAddr: this.queueAddr.toString(),
            withdrawAuthority: this.withdrawAuthority.toString(),
        };
    }
    static fromJSON(obj) {
        return new OracleWalletInitParams({
            oracleAddr: obj.oracleAddr,
            queueAddr: obj.queueAddr,
            withdrawAuthority: obj.withdrawAuthority,
        });
    }
    toMoveStruct() {
        return {
            oracle_addr: this.oracleAddr.toString(),
            queue_addr: this.queueAddr.toString(),
            withdraw_authority: this.withdrawAuthority.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new OracleWalletInitParams({
            oracleAddr: obj.oracle_addr,
            queueAddr: obj.queue_addr,
            withdrawAuthority: obj.withdraw_authority,
        });
    }
}
exports.OracleWalletInitParams = OracleWalletInitParams;
