// @switchboard-xyz/aptos.js/lib/generated/types/OracleWalletContributeParams.js
import { BN } from "@switchboard-xyz/common"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class OracleWalletContributeParams {
    oracleAddr;
    queueAddr;
    loadAmount;
    constructor(fields) {
        this.oracleAddr = fields.oracleAddr;
        this.queueAddr = fields.queueAddr;
        this.loadAmount = fields.loadAmount;
    }
    toJSON() {
        return {
            oracleAddr: this.oracleAddr.toString(),
            queueAddr: this.queueAddr.toString(),
            loadAmount: this.loadAmount.toString(),
        };
    }
    static fromJSON(obj) {
        return new OracleWalletContributeParams({
            oracleAddr: obj.oracleAddr,
            queueAddr: obj.queueAddr,
            loadAmount: new BN(obj.loadAmount),
        });
    }
    toMoveStruct() {
        return {
            oracle_addr: this.oracleAddr.toString(),
            queue_addr: this.queueAddr.toString(),
            load_amount: this.loadAmount.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new OracleWalletContributeParams({
            oracleAddr: obj.oracle_addr,
            queueAddr: obj.queue_addr,
            loadAmount: new BN(obj.load_amount),
        });
    }
}
