// @switchboard-xyz/aptos.js/lib/generated/types/PermissionInitParams.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PermissionInitParams = void 0;
class PermissionInitParams {
    authority;
    granter;
    grantee;
    constructor(fields) {
        this.authority = fields.authority;
        this.granter = fields.granter;
        this.grantee = fields.grantee;
    }
    toJSON() {
        return {
            authority: this.authority.toString(),
            granter: this.granter.toString(),
            grantee: this.grantee.toString(),
        };
    }
    static fromJSON(obj) {
        return new PermissionInitParams({
            authority: obj.authority,
            granter: obj.granter,
            grantee: obj.grantee,
        });
    }
    toMoveStruct() {
        return {
            authority: this.authority.toString(),
            granter: this.granter.toString(),
            grantee: this.grantee.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new PermissionInitParams({
            authority: obj.authority,
            granter: obj.granter,
            grantee: obj.grantee,
        });
    }
}
exports.PermissionInitParams = PermissionInitParams;
