// @switchboard-xyz/aptos.js/lib/generated/types/PermissionSetParams.js
import { BN } from "@switchboard-xyz/common"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class PermissionSetParams {
    authority;
    granter;
    grantee;
    permission;
    enable;
    constructor(fields) {
        this.authority = fields.authority;
        this.granter = fields.granter;
        this.grantee = fields.grantee;
        this.permission = fields.permission;
        this.enable = fields.enable;
    }
    toJSON() {
        return {
            authority: this.authority.toString(),
            granter: this.granter.toString(),
            grantee: this.grantee.toString(),
            permission: this.permission.toString(),
            enable: this.enable,
        };
    }
    static fromJSON(obj) {
        return new PermissionSetParams({
            authority: obj.authority,
            granter: obj.granter,
            grantee: obj.grantee,
            permission: new BN(obj.permission),
            enable: obj.enable,
        });
    }
    toMoveStruct() {
        return {
            authority: this.authority.toString(),
            granter: this.granter.toString(),
            grantee: this.grantee.toString(),
            permission: this.permission.toString(),
            enable: this.enable,
        };
    }
    static fromMoveStruct(obj) {
        return new PermissionSetParams({
            authority: obj.authority,
            granter: obj.granter,
            grantee: obj.grantee,
            permission: new BN(obj.permission),
            enable: obj.enable,
        });
    }
}
