// @switchboard-xyz/aptos.js/lib/generated/types/ProbationBrokenEvent.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProbationBrokenEvent = void 0;
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class ProbationBrokenEvent {
    aggregatorAddress;
    queueAddress;
    timestamp;
    constructor(fields) {
        this.aggregatorAddress = fields.aggregatorAddress;
        this.queueAddress = fields.queueAddress;
        this.timestamp = fields.timestamp;
    }
    toJSON() {
        return {
            aggregatorAddress: this.aggregatorAddress.toString(),
            queueAddress: this.queueAddress.toString(),
            timestamp: this.timestamp.toString(),
        };
    }
    static fromJSON(obj) {
        return new ProbationBrokenEvent({
            aggregatorAddress: obj.aggregatorAddress,
            queueAddress: obj.queueAddress,
            timestamp: new common_1.BN(obj.timestamp),
        });
    }
    toMoveStruct() {
        return {
            aggregator_address: this.aggregatorAddress.toString(),
            queue_address: this.queueAddress.toString(),
            timestamp: this.timestamp.toString(),
        };
    }
    static fromMoveStruct(obj) {
        return new ProbationBrokenEvent({
            aggregatorAddress: obj.aggregator_address,
            queueAddress: obj.queue_address,
            timestamp: new common_1.BN(obj.timestamp),
        });
    }
}
exports.ProbationBrokenEvent = ProbationBrokenEvent;
