// @switchboard-xyz/aptos.js/lib/generated/types/SwitchboardDecimal.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SwitchboardDecimal = void 0;
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class SwitchboardDecimal {
    value;
    dec;
    neg;
    constructor(fields) {
        this.value = fields.value;
        this.dec = fields.dec;
        this.neg = fields.neg;
    }
    toJSON() {
        return {
            value: this.value.toString(),
            dec: this.dec,
            neg: this.neg,
        };
    }
    static fromJSON(obj) {
        return new SwitchboardDecimal({
            value: new common_1.BN(obj.value),
            dec: obj.dec,
            neg: obj.neg,
        });
    }
    toMoveStruct() {
        return {
            value: this.value.toString(),
            dec: this.dec,
            neg: this.neg,
        };
    }
    static fromMoveStruct(obj) {
        return new SwitchboardDecimal({
            value: new common_1.BN(obj.value),
            dec: obj.dec,
            neg: obj.neg,
        });
    }
    toBig() {
        const oldDp = common_1.Big.DP;
        common_1.Big.DP = 18;
        let result = new common_1.Big(this.value.toString());
        if (this.neg === true) {
            result = result.mul(-1);
        }
        const TEN = new common_1.Big(10);
        result = safeDiv(result, TEN.pow(this.dec));
        common_1.Big.DP = oldDp;
        return result;
    }
    static fromBig(val) {
        const value = val.c.slice();
        const e = val.e + 1;
        while (value.length - e > 9) {
            value.pop();
        }
        return new SwitchboardDecimal({
            value: new common_1.BN(value.join("")),
            dec: value.length - e,
            neg: val.s === -1,
        });
    }
}
exports.SwitchboardDecimal = SwitchboardDecimal;
function safeDiv(number_, denominator, decimals = 20) {
    const oldDp = common_1.Big.DP;
    common_1.Big.DP = decimals;
    const result = number_.div(denominator);
    common_1.Big.DP = oldDp;
    return result;
}
