// @switchboard-xyz/aptos.js/lib/SwitchboardProgram.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SwitchboardProgram = exports.getRpcUrl = exports.getProgramId = exports.AptosSimulationError = void 0;
const index_js_1 = require("./generated/index.cjs");
const ts_sdk_1 = require("@aptos-labs/ts-sdk");
class AptosSimulationError extends Error {
    constructor(message) {
        super(`SimulationError: ${message}`);
        Object.setPrototypeOf(this, AptosSimulationError.prototype);
    }
}
exports.AptosSimulationError = AptosSimulationError;
function getProgramId(networkId, programId) {
    if (programId) {
        return programId;
    }
    switch (networkId) {
        case "mainnet":
            return index_js_1.MAINNET_PROGRAM_ID;
        case "testnet":
            return index_js_1.TESTNET_PROGRAM_ID;
        default:
            throw new Error(`Failed to find Aptos ProgramID. Try passing in a programId`);
    }
}
exports.getProgramId = getProgramId;
function getRpcUrl(networkId) {
    switch (networkId) {
        case "mainnet":
            return "https://fullnode.mainnet.aptoslabs.com/v1";
        case "testnet":
            return "https://fullnode.testnet.aptoslabs.com/v1";
        case "devnet":
            return "http://localhost:8080";
        default:
            throw new Error(`Failed to find Aptos RpcUrl for network ${networkId}. Try passing in an rpcUrl`);
    }
}
exports.getRpcUrl = getRpcUrl;
class SwitchboardProgram {
    client;
    network;
    switchboardAddress;
    signer;
    coinType;
    constructor(client, network, switchboardAddress, signer, coinType = "0x1::aptos_coin::AptosCoin") {
        this.client = client;
        this.network = network;
        this.switchboardAddress = switchboardAddress;
        this.signer = signer;
        this.coinType = coinType;
    }
    get READ_ONLY() {
        return this.signer === undefined;
    }
    get newAccount() {
        return ts_sdk_1.Account.generate();
    }
    static getAccount(privateKeyBytes, address) {
        return ts_sdk_1.Account.fromPrivateKey({
            privateKey: new ts_sdk_1.Ed25519PrivateKey(privateKeyBytes),
            address,
        });
    }
    static async load(network, rpcUrl, programId, signer, coinType = "0x1::aptos_coin::AptosCoin") {
        const switchboardAddress = programId ?? getProgramId(network, programId);
        const client = new ts_sdk_1.Aptos(new ts_sdk_1.AptosConfig({
            fullnode: rpcUrl ?? getRpcUrl(network),
            network,
        }));
        return new SwitchboardProgram(client, network, switchboardAddress, signer, coinType);
    }
}
exports.SwitchboardProgram = SwitchboardProgram;
